File: sgd/sgd_lr.m
function [theta] = sgd_lr()
%% SGD for logistic regreesion
% min NLL(w) = min -sum_{i=1}^{n} [y_i log mu_i + (1-y_i) log(1-mu_i)]
% where mu_i = sigm(w'x_i)

clear all; close all; rng('default');

%% synthetic data
n=1e3;

%mu1 = [2,2]; mu2 = [-2,-2];
mu1 = [1,1]; mu2 = [-1,-1];
pik = [0.4, 0.6];

X = zeros(n,2); y = zeros(n,1);
for i=1:n
    u = rand; 
    idx = find(u<cumsum(pik),1,'first');
    if (idx == 1)
        X(i,:) = randn(1,2) + mu1;
        y(i) = 1;
    else
        X(i,:) = randn(1,2) + mu2;
        y(i) = -1;
    end         
end

%figure;
%scatter(X(:,1),X(:,2));

%% SGD for logistic regression

num_iter = 1e2;
theta = randn(size(X,2),1);
lambda = 1e-9; %regularization

tau0 = 10;
kappa = 1; %(0.5,1] 
eta = zeros(num_iter,1); 
for i=1:num_iter
    eta(i) = (tau0 + i)^(-kappa);
end
%figure; plot(eta)
%xlabel('iterations'); ylabel('learning rate \eta_k');
% Leon Bottou, @(t) 1/(lambda*(t+t0)));
% Nic Schraudolph, @(t) eta0*t0/(t+t0));

batchsize = 200;
[batchdata, batchlabels] = make_batches(X, y, batchsize);
num_batches = numel(batchlabels);

num_updates = 0;
J_rec = zeros(num_iter * num_batches,1);
t_rec = zeros(num_iter * num_batches,1);

for iter = 1:num_iter   
    
    for b = 1:num_batches
       Xb = batchdata{b}; yb = batchlabels{b};
       [J_cost, J_grad] = lr_objective(theta, Xb, yb, lambda);
       theta = theta - eta(iter) * (num_batches * J_grad);
       
       num_updates = num_updates + 1;
       J_rec(num_updates) = J_cost;
       t_rec(num_updates) = norm(theta,2);       
    end

    fprintf('iteration %d, cost: %.2f\n',iter,J_cost);
end

y_pred = 2*(sigmoid(X*theta) >= 0.5)-1;
y_err = length(find(y_pred-y))/n;
fprintf('classification error: %.4f\n', y_err);

%% generate plots

figure;
plot(J_rec); title('logistic regression');
xlabel('iterations'); ylabel('cost');

figure;
plot(t_rec); title('logistic regression');
xlabel('iterations'); ylabel('theta l2 norm');

figure;
scatter(X(find(y==1),1),X(find(y==1),2), 'ro'); hold on;
scatter(X(find(y==-1),1),X(find(y==-1),2), 'bo'); hold on;

x1 = linspace(min(X(:,1))-1,max(X(:,1))+1,10);
plot(x1, -(theta(1)/theta(2))*x1, '-r');


end

%% lr objective with l2 penalty
function [cost, grad, H] = lr_objective(theta, X, y, lambda)

n = length(y);
y01 = (y+1)/2;

%compute the objective
mu = sigmoid(X*theta);

mu = max(mu,eps);    %bound away from 0
mu = min(mu,1-eps);  %bound away from 1

cost = -(1/n)*sum(y01.*log(mu)+(1-y01).*log(1-mu)) + sum(lambda.*(theta.^2));

%compute gradient of the lr objective
grad = X'*(mu-y01) + 2*lambda.*theta;

%compute hessian of the lr objective
H = X'*diag(mu.*(1-mu))*X + 2*lambda*eye(length(theta));

end

%% sigmoid function
function [s] = sigmoid(a)
s = 1./(1+exp(-a));
end

%% mini-batch function
function [batchdata, batchlabels] = make_batches(X, y, batchsize)
  n = size(X,1);
  num_batches = ceil(n/batchsize);
  groups = repmat(1:num_batches,1,batchsize);
  groups = groups(1:n);
  batchdata = cell(1, num_batches);
  batchlabels = cell(1, num_batches);
  for i=1:num_batches
    batchdata{i} = X(groups == i,:);
    batchlabels{i} = y(groups == i,:);
  end
end

%% notes

File: sgd/python/sgd_lr.py
import numpy as np
import matplotlib.pyplot as plt
import time
from sklearn.datasets import load_iris

np.random.seed(0)

class sgdlr:
    
    def __init__(self):
        
        self.num_iter = 100
        self.lmbda = 1e-9
        
        self.tau0 = 10
        self.kappa = 1
        
        self.batchsize = 200
        
    def fit(self,X,y):

        theta = np.random.randn(X.shape[1],1)
        #learning rate
        eta = np.zeros((self.num_iter,1))
        for i in range(self.num_iter):
            eta[i] = (self.tau0+i)**(-self.kappa)        
                
        #divide data in batches        
        batchdata, batchlabels = self.make_batches(X,y,self.batchsize)                    
        num_batches = batchdata.shape[0]
        num_updates = 0
        
        J_rec = np.zeros((self.num_iter * num_batches,1))
        t_rec = np.zeros((self.num_iter * num_batches,1))
        
        for itr in range(self.num_iter):
            for b in range(num_batches):
                Xb = batchdata[b]
                yb = batchlabels[b]
                J_cost, J_grad = self.lr_objective(theta, Xb, yb, self.lmbda)
                theta = theta - eta[itr]*(num_batches*J_grad)                
                J_rec[num_updates] = J_cost
                t_rec[num_updates] = np.linalg.norm(theta,2)
                num_updates = num_updates+1
            print "iteration %d, cost: %f" %(itr, J_cost)
        
        y_pred = 2*(self.sigmoid(X.dot(theta)) > 0.5) - 1
        y_err = np.size(np.where(y_pred - y)[0])/float(y.shape[0])
        print "classification error: %f" %(y_err)        

        self.generate_plots(X, J_rec, t_rec, theta)
        return theta        
    
    def make_batches(self,X,y,batchsize):
        n = X.shape[0]
        d = X.shape[1]
        num_batches = int(np.ceil(n/batchsize))
        groups = np.tile(range(num_batches),batchsize)
        batchdata=np.zeros((num_batches,batchsize,d))
        batchlabels=np.zeros((num_batches,batchsize,1))
        for i in range(num_batches):
            batchdata[i,:,:] = X[groups==i,:]
            batchlabels[i,:] = y[groups==i]
            
        return batchdata, batchlabels
            
    def lr_objective(self,theta,X,y,lmbda):
        
        n = y.shape[0]
        y01 = (y+1)/2.0
        
        #compute the objective
        mu = self.sigmoid(X.dot(theta))
    
        #bound away from 0 and 1
        eps = np.finfo(float).eps   
        mu = np.maximum(mu,eps)
        mu = np.minimum(mu,1-eps)    
        
        #compute cost
        cost = -(1/n)*np.sum(y01*np.log(mu)+(1-y01)*np.log(1-mu))+np.sum(lmbda*theta*theta)
        
        #compute the gradient of the lr objective
        grad = X.T.dot(mu-y01) + 2*lmbda*theta
        
        #compute the Hessian of the lr objective
        #H = X.T.dot(np.diag(np.diag( mu*(1-mu) ))).dot(X) + 2*lmbda*np.eye(np.size(theta))
        
        return cost, grad
        
                    
    def sigmoid(self,a):
        return 1/(1+np.exp(-a))
    
    def generate_plots(self,X,J_rec,t_rec,theta):
                        
        plt.figure()
        plt.plot(J_rec)
        plt.title("logistic regression")
        plt.xlabel('iterations')
        plt.ylabel('cost')
        plt.show()
        
        plt.figure()
        plt.plot(t_rec)
        plt.title("logistic regression")
        plt.xlabel('iterations')
        plt.ylabel('theta l2 norm')
        plt.show()                
        
        plt.figure()
        x1 = np.linspace(np.min(X[:,0])-1,np.max(X[:,0])+1,10)
        plt.scatter(X[:,0], X[:,1])
        plt.plot(x1, -(theta[0]/theta[1])*x1)
        plt.title('logistic regression')
        plt.grid(True)
        plt.xlabel('X1')
        plt.ylabel('X2')
        plt.show()
        
    
    def generate_data(self):
        
        n = 1000
        mu1 = np.array([1,1])
        mu2 = np.array([-1,-1])        
        pik = np.array([0.4,0.6])
        
        X = np.zeros((n,2))
        y = np.zeros((n,1))
        
        for i in range(1,n):
            u = np.random.rand()
            idx = np.where(u < np.cumsum(pik))[0]                            
            
            if (len(idx)==1):
                X[i,:] = np.random.randn(1,2) + mu1
                y[i] = 1
            else:
                X[i,:] = np.random.randn(1,2) + mu2
                y[i] = -1
                
        return X, y    
        
                

if __name__ == "__main__":
    
    #iris = load_iris()
    #X = iris.data
    #y = iris.target
    
    sgd = sgdlr()
    X, y = sgd.generate_data()
    theta = sgd.fit(X,y)
    
    